lib/Zonemaster/Backend/Validator.pm

```diff
@@ -7,36 +7,83 @@ use warnings;
+  untaint_mariadb_database
+  untaint_mariadb_user
+  untaint_password
+  untaint_positive_int
+  untaint_positive_millis
+  untaint_postgresql_ident
+  untaint_unsigned_int
 );
 
 our %EXPORT_TAGS = (
     untaint => [
         qw(
+          untaint_abs_path
+          untaint_ldh_domain
           untaint_engine_type
+          untaint_mariadb_database
+          untaint_mariadb_user
+          untaint_password
+          untaint_positive_int
+          untaint_positive_millis
+          untaint_postgresql_ident
+          untaint_unsigned_int
           )
     ],
 );
 
+# Does not check value ranges within the groups
 Readonly my $IPV4_RE => qr/^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/;
+
+# Does not check the length and number of the hex groups, nor the value ranges in the IPv4 groups
 Readonly my $IPV6_RE => qr/^([0-9a-f]{1,4}:[0-9a-f:]{1,}(:[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})?)$|([0-9a-f]{1,4}::[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})$/i;
 
-Readonly my $API_KEY_RE             => qr/^[a-z0-9-_]{1,512}$/i;
-Readonly my $CLIENT_ID_RE           => qr/^[a-z0-9-+~_.: ]{1,50}$/i;
-Readonly my $CLIENT_VERSION_RE      => qr/^[a-z0-9-+~_.: ]{1,50}$/i;
-Readonly my $DIGEST_RE              => qr/^[a-f0-9]{40}$|^[a-f0-9]{64}$/i;
-Readonly my $ENGINE_TYPE_RE         => qr/^(?:mysql|postgresql|sqlite)$/i;
-Readonly my $IPADDR_RE              => qr/^$|$IPV4_RE|$IPV6_RE/;
-Readonly my $JSONRPC_METHOD_RE      => qr/^[a-z0-9_-]*$/i;
-Readonly my $LANGUAGE_RE            => qr/^[a-z]{2}(_[A-Z]{2})?$/;
+Readonly my $API_KEY_RE                 => qr/^[a-z0-9-_]{1,512}$/i;
+Readonly my $CLIENT_ID_RE               => qr/^[a-z0-9-+~_.: ]{1,50}$/i;
+Readonly my $CLIENT_VERSION_RE          => qr/^[a-z0-9-+~_.: ]{1,50}$/i;
+Readonly my $DIGEST_RE                  => qr/^[a-f0-9]{40}$|^[a-f0-9]{64}$/i;
+Readonly my $ENGINE_TYPE_RE             => qr/^(?:mysql|postgresql|sqlite)$/i;
+Readonly my $IPADDR_RE                  => qr/^$|$IPV4_RE|$IPV6_RE/;
+Readonly my $JSONRPC_METHOD_RE          => qr/^[a-z0-9_-]*$/i;
+Readonly my $LANGUAGE_RE                => qr/^[a-z]{2}(_[A-Z]{2})?$/;
+Readonly my $LDH_DOMAIN_RE1             => qr{^[a-z0-9-.]{1,253}[.]?$}i;
+Readonly my $LDH_DOMAIN_RE2             => qr{^(?:[.]|[^.]{1,63}(?:[.][^.]{1,63})*[.]?)$};
+Readonly my $MARIADB_DATABASE_LENGTH_RE => qr/^.{1,64}$/;
+
+# See: https://mariadb.com/kb/en/identifier-names/#unquoted
+Readonly my $MARIADB_IDENT_RE       => qr/^[0-9a-z\$_]+$/i;
+Readonly my $MARIADB_USER_LENGTH_RE => qr/^.{1,80}$/u;
+
+# Up to 5 and 3 digits in the integer and fraction components respectively
+Readonly my $MILLIS_RE => qr/^(?:0|[1-9][0-9]{0,4})(?:[.][0-9]{1,3})?$/;
+
+# Printable ASCII but first character must not be space or '<'
+Readonly my $PASSWORD_RE => qr/^(?:[\x21-\x3b\x3d-\x7e][\x20-\x7e]{0,99})?$/;
+
+# Allow for 5 significant digits
+Readonly my $POSITIVE_INT_RE => qr/^[1-9][0-9]{0,4}$/;
+
+# At least one non-zero digit
+Readonly my $POSITIVE_NUM_RE => qr/[^0.]/;
```

Comment:
Not sure if this does what the comment says...

And looking at the code, this is only used in [`untaint_positive_millis`](https://github.com/zonemaster/zonemaster-backend/blob/bcfed7d49ab211589bcb6d3183d85e2352db575b/lib/Zonemaster/Backend/Validator.pm#L218..L221):

```
sub untaint_positive_millis {
    my ( $value ) = @_;
    return _untaint_pat( $value,  $MILLIS_RE, $POSITIVE_NUM_RE );
}
```
And here is the definition of `$MILLIS_RE` :
```
Readonly my $MILLIS_RE => qr/^(?:0|[1-9][0-9]{0,4})(?:[.][0-9]{1,3})?$/;
```

This looks like the `MILLIS_RE` regex is enough to match a positive decimal number between `0` and `99999.999`

(TBH I haven't tried with the unit tests in `t/validator.t` yet)

Reply:
Looking at the unit tests in `t/validator.t` I understand that we refuse null values. So we do need something more to avoid the `0`. Would it be possible to make it clearer in the method's name (for instance `untaint_strictly_positive_millis`) ?
And I think the `POSITIVE_NUM_RE` can be changed to `[1-9]` so that it reflects what the comment says.

Reply:
I replaced the POSITIVE_NUM_RE with a NON_ZERO_NUM that uses `[1-9]`, per your suggestion. I think the resulting code is much better. I used that in untaint_positive_int() as well together with UNSIGNED_INT_RE, and this made it possible to delete POSITIVE_INT_RE.

Regarding the membership of zero in the set of positive numbers, see my comment below.